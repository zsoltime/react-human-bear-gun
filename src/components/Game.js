import React, { useEffect, useState } from 'react';
import random from 'utils.random';

import RadioBox from './RadioBox';
import Results from './Results';
import Icon from './Icons';

const getResult = (userChoice, computerChoice) => {
  if (userChoice === computerChoice) {
    return { result: "It's a tie. Try again!", winner: null };
  }

  switch (userChoice) {
    case 'Human':
      return computerChoice === 'Bear'
        ? {
            result: "You've just been mauled by a bear",
            winner: 'computer',
          }
        : {
            result: "You've just disarmed your opponent",
            winner: 'user',
          };

    case 'Bear':
      return computerChoice === 'Human'
        ? {
            result: 'You tore him apart like a loaf of bread',
            winner: 'user',
          }
        : {
            result: 'Oops, you have been shot by a gun',
            winner: 'computer',
          };

    case 'Gun':
      return computerChoice === 'Human'
        ? {
            result: "It looks like you've been disarmed",
            winner: 'computer',
          }
        : { result: 'Geez, you shot the bear', winner: 'user' };

    default:
      return {
        result: null,
        winner: null,
      };
  }
};

const Game = () => {
  const CHOICES = ['Human', 'Bear', 'Gun'];
  const STORAGE_NAME = 'humanBearGunPoints';
  const [computerChoice, setComputerChoice] = useState(null);
  const [computerIsThinking, setComputerIsThinking] = useState(false);
  const [userChoice, setUserChoice] = useState(null);
  const [points, setPoints] = useState({
    computer: 0,
    user: 0,
  });
  const [result, setResult] = useState(null);
  const [winner, setWinner] = useState(null);

  useEffect(() => {
    const savePoints = () => {
      localStorage.setItem(STORAGE_NAME, JSON.stringify(points));
    };

    window.addEventListener('beforeunload', savePoints);

    return () => {
      window.removeEventListener('beforeunload', savePoints);
    };
  }, [points]);

  useEffect(() => {
    try {
      const pointsFromStorage = JSON.parse(
        window.localStorage.getItem(STORAGE_NAME)
      );
      if (pointsFromStorage) {
        setPoints(pointsFromStorage);
      }
    } catch (e) {
      console.error(e);
    }
  }, []);

  const markUserChoice = index => {
    setComputerChoice(null);
    setUserChoice(CHOICES[index]);
    setResult(null);
    setWinner(null);
  };

  const playGame = e => {
    e.preventDefault();

    const randomResult = random(CHOICES);
    const response = getResult(userChoice, randomResult);

    setComputerChoice(null);
    setComputerIsThinking(true);
    setResult(null);
    setWinner(null);

    window.setTimeout(() => {
      setComputerChoice(randomResult);
      setComputerIsThinking(false);
      setResult(response.result);
      setWinner(response.winner);
      setPoints(prevPoints => ({
        ...prevPoints,
        [response.winner]: prevPoints[response.winner] + 1,
      }));
    }, random(300, 1000));
  };

  const resetGame = () => {
    setComputerChoice(null);
    setComputerIsThinking(false);
    setUserChoice(null);
    setResult(null);
    setPoints({
      computer: 0,
      user: 0,
    });
    setWinner(null);
  };

  return (
    <div className="wrapper">
      <h1 className="title">
        Human-Bear-Gun <strong>Mini Game</strong>
      </h1>
      <form className="form" onSubmit={playGame}>
        {/* Can't use fieldset and legend as display: flex doesn't work in Chrome :( */}
        <div
          className="game"
          role="radiogroup"
          aria-labelledby="label"
        >
          <div className="instructions" id="label">
            {`Select one of the ${CHOICES.length} choices`}
          </div>
          {CHOICES.map((choice, i) => (
            <RadioBox
              key={choice}
              choice={choice}
              userChoice={userChoice}
              onChange={() => markUserChoice(i)}
            />
          ))}
        </div>
        <button
          className="btn"
          disabled={!userChoice || computerIsThinking}
        >
          {computerIsThinking
            ? 'Your opponent is thinking...'
            : 'Play'}
        </button>
      </form>
      <Results
        computerChoice={computerChoice}
        userChoice={userChoice}
        points={points}
        result={result}
        winner={winner}
      />
      <button
        className="btn btn--reset"
        type="button"
        title="Reset game"
        onClick={resetGame}
      >
        <Icon type="Reset" size="24px" />
        <span className="visuallyhidden">Reset Game</span>
      </button>
    </div>
  );
};

export default Game;
