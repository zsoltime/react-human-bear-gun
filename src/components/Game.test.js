import React from 'react';
import { shallow } from 'enzyme';

import Game, { getResult } from './Game';

describe('<Game />', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<Game />);

    expect(wrapper).toMatchSnapshot();
  });

  // TODO: More tests :)
});

describe('getResult()', () => {
  it('returns the correct winner', () => {
    const CHOICES = ['Human', 'Bear', 'Gun'];
    const algo = [
      { user: CHOICES[0], computer: CHOICES[0], winner: null },
      { user: CHOICES[1], computer: CHOICES[0], winner: 'user' },
      { user: CHOICES[2], computer: CHOICES[0], winner: 'computer' },
      { user: CHOICES[0], computer: CHOICES[1], winner: 'computer' },
      { user: CHOICES[1], computer: CHOICES[1], winner: null },
      { user: CHOICES[2], computer: CHOICES[1], winner: 'user' },
      { user: CHOICES[0], computer: CHOICES[2], winner: 'user' },
      { user: CHOICES[1], computer: CHOICES[2], winner: 'computer' },
      { user: CHOICES[2], computer: CHOICES[2], winner: null },
    ];

    for (let i = 0; i < algo.length; i++) {
      const result = getResult(algo[i].user, algo[i].computer);

      expect(result.winner).toBe(algo[i].winner);
    }
  });
});
