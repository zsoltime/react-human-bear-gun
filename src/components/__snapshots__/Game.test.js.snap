// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> matches snapshot 1`] = `
<div
  className="wrapper"
>
  <h1
    className="title"
  >
    Human-Bear-Gun 
    <strong>
      Mini Game
    </strong>
  </h1>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      aria-labelledby="label"
      className="game"
      role="radiogroup"
    >
      <div
        className="instructions"
        id="label"
      >
        Select one of the 3 choices
      </div>
      <RadioBox
        choice="Human"
        key="Human"
        onChange={[Function]}
        userChoice={null}
      />
      <RadioBox
        choice="Bear"
        key="Bear"
        onChange={[Function]}
        userChoice={null}
      />
      <RadioBox
        choice="Gun"
        key="Gun"
        onChange={[Function]}
        userChoice={null}
      />
    </div>
    <button
      className="btn"
      disabled={true}
    >
      Play
    </button>
  </form>
  <Results
    computerChoice={null}
    points={
      Object {
        "computer": 0,
        "user": 0,
      }
    }
    result={null}
    userChoice={null}
    winner={null}
  />
  <button
    className="btn btn--reset"
    onClick={[Function]}
    title="Reset game"
    type="button"
  >
    <Icon
      size="24px"
      type="Reset"
    />
    <span
      className="visuallyhidden"
    >
      Reset Game
    </span>
  </button>
</div>
`;
